pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,325 @@
+func (q *Queue) CanAdd(ids ...string) error {
+	var err error
+	for _, id := range ids {
+		if _, ok := q.candidateProviderIDToCommandID[id]; ok {
+			err = multierr.Append(err, fmt.Errorf("candidate is part of active deprovisioning decision"))
+		}
+	}
+	return err
+}
+
+// Handle will check for timeouts, then execute a wait or termination.
+// If there
+func (q *Queue) Handle(ctx context.Context, cmd *Command) (bool, error) {
+	// logging.FromContext(ctx).Infof("DEBUGGING: This is the current time: %s, this is the time added: %s", q.clock)
+	if q.clock.Since(cmd.TimeAdded) > maxRetryDuration {
+		return false, fmt.Errorf("abandoning command, reached timeout, %w", cmd.LastError)
+	}
+	// If the time hasn't expired, either wait, or terminate.
+	requeue, err := q.WaitOrTerminate(ctx, cmd)
+	if err != nil {
+		// If there was an error, set this as the command's last error so that we can propagate it.
+		cmd.LastError = err
+		if !retry.IsRecoverable(err) {
+			return false, nil
+		}
+		return requeue, nil
+	}
+	return requeue, nil
+}
+
+// WaitOrTerminate will wait until launched machines are ready.
+// Once the replacements are ready, it will terminate the candidates.
+// Will return true if the item in the queue should be re-queued. If a command has
+// timed out, this will return false.
+// nolint:gocyclo
+func (q *Queue) WaitOrTerminate(ctx context.Context, cmd *Command) (bool, error) {
+	waitErrs := make([]error, len(cmd.ReplacementKeys))
+	for i := range cmd.ReplacementKeys {
+		key := cmd.ReplacementKeys[i]
+		// If we know the node claim is initialized, no need to check again.
+		if key.Initialized {
+			continue
+		}
+		// Get the nodeclaim
+		nodeClaim, err := nodeclaim.Get(ctx, q.kubeClient, key.Key)
+		if err != nil {
+			// The NodeClaim got deleted after an initial eventual consistency delay
+			// This means that there was an ICE error or the Node initializationTTL expired
+			// In this case, the error is unrecoverable, so don't requeue.
+			if errors.IsNotFound(err) && q.clock.Since(cmd.TimeAdded) > time.Second*5 {
+				return false, retry.Unrecoverable(fmt.Errorf("getting machine, %w", err))
+			}
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		// We emitted this event when Deprovisioning was blocked on launching/termination.
+		// This does not block other forms of deprovisioning, but we should still emit this.
+		q.recorder.Publish(deprovisioningevents.Launching(nodeClaim, cmd.Reason))
+		if !nodeClaim.StatusConditions().GetCondition(v1beta1.Initialized).IsTrue() {
+			q.recorder.Publish(deprovisioningevents.WaitingOnReadiness(nodeClaim))
+			waitErrs[i] = fmt.Errorf("getting node claim, %w", err)
+			continue
+		}
+		key.Initialized = true
+		// This should only be reached once since initialization is checked at the beginning.
+		deprovisioningReplacementNodeInitializedHistogram.Observe(q.clock.Since(cmd.TimeAdded).Seconds())
```

Comment:
musing: Is there value in this being a deprovisioning node initilized histogram vs. just a standard launch? I'm thinking no so I'm hoping that we might be able to pull this out in favor of a general initialized histogram metric here as we prepare for v1.

Reply:
Definitely think this makes more sense in the provisioning code/another controller long term 👍